File: solutions/ruby/bob/2/bob.rb
#!/usr/bin/env ruby -w

class Bob
  def hey(phrase)
    text_factory.create(phrase).response
  end
end

class TextFactory
  def initialize
    @types = []
  end

  def create(text)
    @types.each { |text_type| return text_type if text_type.is(text) }
  end

  def register(type)
    @types << type
  end
end

def text_factory
  $factory ||= TextFactory.new
end

class Silence
  def is(text)
    text.strip.empty?
  end

  def response
    'Fine. Be that way!'
  end
end

class Scream
  def is(text)
    text =~ /\w{2}/ && text == text.upcase
  end

  def response
    'Whoa, chill out!'
  end
end

class Question
  def is(text)
    text.end_with?('?')
  end

  def response
    'Sure.'
  end
end

class NoneOfTheAbove
  def is(_)
    true
  end

  def response
    'Whatever.'
  end
end

text_factory.register Silence.new
text_factory.register Scream.new
text_factory.register Question.new
text_factory.register NoneOfTheAbove.new

File: solutions/zig/bob/2/bob.zig
const std = @import("std");
const ascii = std.ascii;
const mem = std.mem;

fn question(input: []const u8) bool {
    return input[input.len - 1] == '?';
}

fn shout(input: []const u8) bool {
    var containsUpper = false;
    for (input) |c| {
        if (ascii.isUpper(c)) containsUpper = true;
        if (ascii.isAlphabetic(c) and ascii.isLower(c)) return false;
    }
    return containsUpper;
}

pub fn response(s: []const u8) []const u8 {
    const t = mem.trim(u8, s, &ascii.whitespace);

    if (t.len == 0) return "Fine. Be that way!";
    if (question(t) and shout(t)) return "Calm down, I know what I'm doing!";
    if (question(t)) return "Sure.";
    if (shout(t)) return "Whoa, chill out!";
    return "Whatever.";
}
